openff/toolkit/topology/molecule.py

```diff
@@ -1971,23 +1982,15 @@ def from_smiles(
         return molecule
 
     def _is_exactly_the_same_as(self, other):
-        for atom1, atom2 in zip(self.atoms, other.atoms):
-            if (
-                (atom1.atomic_number != atom2.atomic_number)
-                or (atom1.formal_charge != atom2.formal_charge)
-                or (atom1.is_aromatic != atom2.is_aromatic)
-                or (atom1.stereochemistry != atom2.stereochemistry)
-            ):
-                return False
-        for bond1, bond2 in zip(self.bonds, other.bonds):
-            if (
-                (bond1.atom1_index != bond2.atom1_index)
-                or (bond1.atom2_index != bond2.atom2_index)
-                or (bond1.is_aromatic != bond2.is_aromatic)
-                or (bond1.stereochemistry != bond2.stereochemistry)
-            ):
-                return False
-        return True
+        self_id = (
```

Comment:
```suggestion
        # Pre-assign molecule atom indices in O(N) time to avoid use of List.index to get index of each one
        # in O(N^2) time
        for index, atom in enumerate(self.atoms):
            atom._molecule_atom_index = index
        for index, atom in enumerate(other.atoms):
            atom._molecule_atom_index = index

        self_id = (
```

Sorry for burying this point earlier - My finding was that the majority of the speedup from using the hash function comes from pre-assigning molecule atom indices, and not from the actual hash comparison. To get the speedup, they need to be somewhere - I think when I suggested the tuple comparison it was in the context of moving them to `_add_atom()`. So if that's not happening, this needs to be here.

Reply:
Ah, gotcha. Thanks for the clarification!